Automate the release process, produce pre-compiled binaries

Veetaha opened the pull request:
Closes #193

This PR adds two new CI workflows `release` and `release-on-tag`.
The release process and artifacts are described in [`docs/internal/release.md`](https://github.com/Veetaha/marker/blob/feat/automate-releases/docs/internal/release.md) in this PR,
so I won't repeat it here.

You can see how it works in my fork. 
Here is an example `release` worklfow run ([link](https://github.com/Veetaha/marker/actions/runs/6137532417/job/16653466213)).
Here is an example `release-on-tag` workflow run ([link](https://github.com/Veetaha/marker/actions/runs/6137533185/job/16653467821)) (the `crates-io` job is failing becase I don't have access to publishing marker crates).

Here are the two generated commits ([link](https://github.com/Veetaha/marker/commits/master)):
![image](https://github.com/rust-marker/marker/assets/36276403/6cd0d225-8cb2-4670-9017-443169e2b6b2)

Here is the github release that it makes ([link](https://github.com/Veetaha/marker/releases)):
![image](https://github.com/rust-marker/marker/assets/36276403/0ef81ea1-2d9b-4761-a525-705241ba60dc)

The other change made in this PR is the addition of `package.rust_version`
property to `Cargo.toml` of the library crates `marker_api`, `marker_uitest`,
`marker_utils`.

I was thinking to refactor the layout of the crates in the repo to separate
the library crates that have a strict MSRV, and the pre-compiled crates that
depend on the pinned nightly toolchain version (https://github.com/rust-marker/marker/issues/193#issuecomment-1703792584),
but I decided not to do that for now. It's not a big deal, and I don't want to
extend this PR with even more changes for that.

The scope of this PR doesn't include the code changes in `cargo-marker` to
make it use the pre-compiled `marker_rustc_driver` from the GitHub release artifacts,
falling back to `cargo install` if that doesn't work. That should be implemented in
the scope of #238.

---

The automation also covers publishing crates to crates.io with `cargo-release`

As a result of this PR the following changes to `marker` infrastructure need to be done manually and only once.

The new `bash-test` CI job needs to be marked as required to pass in `master` branch protection rules in the repo settings.
This test validates that the edits that the workflow makes to the repo match the expected git diff snapshot.

The `set-version.sh` script does a bunch of regex replacements with `sed` in all `.md` files and in some `.rs` files to update the places where we mention the bare version numbers. For example the constants of marker API and driver versions in the `cargo-marker` code, and the references to `marker_lints/api/uitest = "x.y.z"` all over the documentation.

---
The new `RELEASE_PAT` secret needs to be created in the repository. This is required because the default `github.token` that the workflows get doesn't work for the tag-event-based flow. See [the comment here](https://github.com/Veetaha/marker/blob/ac0ffbfbcfb75a68cabf75033dd9e48cbf95b46c/.github/workflows/release.yml#L25-L29).

To create it follow this link: https://github.com/settings/personal-access-tokens/new.

Make a token with a really long expiration date (unless you are okay with having a shorter expiration duration and rotating the token manually on schedule). For the permissions allow access only to the marker repository and give it "read and write" for the repo contents.

<details>
<summary>Screenshots</summary>

![](https://github.com/toml-lang/toml/assets/36276403/88084992-3ac2-4592-b31d-1cf1de3d2afb)
![](https://github.com/toml-lang/toml/assets/36276403/23fb6808-0f16-4b25-a5db-96c961c61e2a)

</details>

Then add this as as a GitHub Actions secret in the repo settings.

<details>
<summary>Screenshot</summary>

![](https://github.com/toml-lang/toml/assets/36276403/a3d83391-8373-4b08-baa9-b25a391f09d9)

</details>

---

Generate a new crates.io API token by following this link: https://crates.io/settings/tokens.
The token needs permissions to `publish-update` and optionally to `publish-new`. Unless you are going to manually reserve the crate names, and not rely on the release CI to publish the first version of the crate you may not give it `publish-new` permission.

The same comment about the token expiration applies here as well. Unfortunatelly, there isn't an easy way to automate the rotation of the expired tokens, so you either have to do that manually which is secure but annoying, or you can set a really high expiration duration, which is convenient but unsecure.

You can also restrict the token with the crate name pattern like `marker_*`.

The token then needs to be set as a secret named `CRATES_IO_TOKEN` in the repository. On my screenshots above I didn't have that secret, and I actually didn't fully test the publishing of the crates to crates.io because I don't have access to `marker` crates on crates.io, and otherwise I wouldn't like to create junk crates just for testing that will be left on `crates.io` indefinitely. I expect there may be publishing failures on the first release that uses this automation because I didn't test this entirely, but hopefully, there won't be any, or they will be simple to fix.

Veetaha commented on marker_api/Cargo.toml:
```diff
@@ -4,19 +4,18 @@ name = "marker_api"
+version      = { workspace = true }
 
 [dependencies]
-visibility = { workspace = true }
```
With this change `marker_api` for lint crates now hase zero dependencies :tada:

Veetaha commented on marker_api/Cargo.toml:
```diff
@@ -4,19 +4,18 @@ name = "marker_api"
 
 [dependencies]
-visibility = { workspace = true }
+visibility = { optional = true, workspace = true }
```
With this change `marker_api` for lint crates now has zero dependencies :tada:

xFrednet reviewed (commented):
I've skimmed over the documentation and some files. I'll continue the review during this week.

Thank you so much for this PR and also all the documentation you wrote. This must have taken a lot of effort and I highly appreciate it!

xFrednet commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,61 @@
+      # Kudos to this guy for sharing how to set the Github Actions bot identity:
+      # https://github.com/actions/checkout/issues/13#issuecomment-724415212
+      - run: |
+          git config user.name "Github Actions"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
```
I've created @rust-marker-ci to create the tokens. We can just use these user values. It feels less hacky ^^

```suggestion
          git config user.name "rust-marker-ci"
          git config user.email "rust.marker.ci@gmail.com"
```

Though, I liked the hack. :D

xFrednet commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,61 @@
+          # By design, GitHub doesn't trigger workflows on pushes to tags created by
+          # the `github.token` to prevent infinite loops.
+          # https://github.com/orgs/community/discussions/27028
+          token: ${{ secrets.RELEASE_PAT }}
```
Is there a reason why this is called `RELEASE_PAT` and not `RELEASE_TOKEN` or something similar?

xFrednet commented on scripts/release/qemu-install.sh:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
```
Could you add a doc comment to this script to explain a bit what it does? I feel like I understand it now, but I think a small comment at the beginning would make this clearer, also why this installs rust's toolchain using rustup here :)

xFrednet reviewed (commented):
I was finally able to read through all your changes, and I want to say it one more: Thank you!

It all looks clean and is also well documented, while being a fun read :D This is how reviewing can be fun! I'm excited to test this with the next release :)

Veetaha commented on scripts/release/set-version.sh:
```diff
@@ -0,0 +1,85 @@
+with_log cargo update --workspace
+
+if [[ $commit != "" ]]; then
+    git add --all -- :!rust-toolchain.toml
```
I forgot to leave a comment here, but this essentially means "commit all changed files, but ignore changes to rust-toolchain.toml. This is because on release.yml CI we remove it

Veetaha commented on scripts/download/taplo.sh:
```diff
@@ -14,5 +14,5 @@ file_stem=taplo-$os-$arch_rust
-with_log chmod +x $file_stem$exe
-with_log mv $file_stem$exe ~/.cargo/bin/taplo$exe
+with_log chmod +x $file_stem
+with_log mv $file_stem ~/.cargo/bin/taplo
```
Missed this, thanks, corrected.

Veetaha commented:
@xFrednet I addressed all your review comments. However, I mistakenly force-pushed the branch. You may review only the changes  starting with the commit "Address xFrednet and self-review".

In these changes I additionally added support for releasing versions with an arbitrary `-[a-zA-Z0-9.]` suffix. Meaning with this automation it will be possible to release say `v1.0.0-rc` or `v1.0.0-rc.2` pre-release version. Release of such a version will update only the Cargo.toml and `driver.rs` file versions. All the documentation will still refer to the stable version.

The sections `region replace-version stable/unstable` specify wether we want to have the latest stable marker version mentioned in the region, or the unstable version.

xFrednet commented:
You'll get a full review in the coming days. For now, I have to say that I absolutely love the region markers, to replace the Marker version. This will make it so much easier to do releases! It really feels like v0.3 will be the greatest releases to date :D

xFrednet commented on docs/internal/release.md:
```diff
@@ -0,0 +1,111 @@
+
+GitHub is very generous with the releases storage limits. They are [almost unlimited](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases#storage-and-bandwidth-quotas). They difinitely want your project to be open-source 😉.
+
+The `sha256` sum is a small file that users may optionally download together with the archive itself to verify the integrity of the archive. It serves as a signature of the artifact to make sure it was downloaded as expected bit-by-bit with what was published effectively detecting corruptions during the download and making it harder to forge the artifact for the malicious actor.
```
```suggestion
The `sha256` sum is a small file that users may optionally download together with the archive itself to verify the integrity of the archive. It serves as a signature of the artifact to make sure it was downloaded as expected bit-by-bit with what was published effectively detecting corruptions during the download and making it harder to forge artifacts for malicious actors.
```

xFrednet commented on docs/internal/release.md:
````diff
@@ -0,0 +1,111 @@
+
+The `sha256` sum is a small file that users may optionally download together with the archive itself to verify the integrity of the archive. It serves as a signature of the artifact to make sure it was downloaded as expected bit-by-bit with what was published effectively detecting corruptions during the download and making it harder to forge the artifact for the malicious actor.
+
+```bash
````
I find this script a bit weird, it seems like an abbreviated version of the installation script. It might be better to just link to that one, even if this is just internal documentation.

Instead, we could potentially write:

> <-- region replace-version unstable -->
> This [`install.sh`](https://raw.githubusercontent.com/Veetaha/marker/v0.2.1/install.sh) script, can be used to automatically download and verify Marker's binaries.
> <-- endregion replace-version unstable -->

If we want to keep it, there should be region markers for the version:
```md
<!-- region replace-version unstable -->
...
<!-- endregion replace-version unstable -->
```

Veetaha replied:
Agreed, replaced it with the reference to the script.

xFrednet reviewed (approved):
This version looks good to me, I have some final comments, and then we can merge this! :tada: 

Would you mind squashing some of your commits? You can just force-push. Github can provide diffs, if the branch was not rebased on another one :)

Veetaha commented on install.sh:
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env bash
```
Now that you mention it, I don't like that this script mentions a `-dev` version, because it simply won't ever work, because we'll never actually make a `-dev` release.

Maybe the the `replace-version` regions should have three states:
- `stable` (will be set to the latest suffixless version)
- `unstable` (will be set to the latest released version, since `-dev` isn't released, it won't include it, but it will include things like `1.0.0-rc`)
- `dev` (will be set any latest version, including `-dev`)

xFrednet replied:
Having three states for this would make a lot of sense. Are you okay with adding the third state? It's totally fine, if you want to be done, then we can merge it as is. :)

Thank you for moving the file! :D

Veetaha commented on README.md:
```diff
@@ -35,21 +35,41 @@ And more to come, see Marker's goals and limitations below.
 
 ### Installation
 
+#### Download pre-compiled binaries (recommended)
```
I may leave these sections commented out for now, but don't forget to uncomment them right before the 0.3.0 release

Veetaha replied:
I was thinking of commenting out these sections, but that looks quite ugly because you can't nest comments in markdown. Anyway, I'll move them to a separate PR

Veetaha replied:
Actually, making a PR doesn't make sense right now before this PR is merged, especially because there will be a merge commit created.

You can use `git revert 0f17cd9` before the release to revert the commit that removes the docs for the pre-compiled binaries

xFrednet replied:
Sounds like a perfect solution! :D

Veetaha commented on README.md:
```diff
@@ -35,21 +35,41 @@ And more to come, see Marker's goals and limitations below.
+
+We provide pre-compiled binaries for the mainstream platforms. See the list of available artifacts in our [Github Releases](https://github.com/rust-marker/marker/releases/latest).
+
+Run the following bash script to install the required rust toolchain dependency on your machine and download the current version of `cargo-marker` CLI, and the internal driver.
```
I also like having this in the readme, I think it's fine to keep them in sync. I bet there is a tool that allows synchronizing regions of text somewhere out there similar to our custom `replace-version`, but I hope this problem will be a rare occasion.

Actually, I would really like to have windows support for the install script. I just didn't make that effort due to laziness, powershell is even worse than bash because I don't ever write it. I wish we could use a cross-platform shell like nutshell, but that one isn't at `1.0` yet, and isn't installed by default on windows and UNIX, unfortunately.

I'll try to add a PowerShell install script impl, but in a separate PR, I'd like to merge this one without that to not make "the ideal" the enemy of "the good". But for now, I'll add that clarification

Veetaha commented:
I addressed the review comments, and did some more improvements.

I removed `--retry-delay` from curl. I just noticed that it disables the exponential backoff retry ([StackOverflow](https://stackoverflow.com/a/49825264/9259330)), which is unwanted; exponential backoff is better than a fixed one because this way we don't overload github if it is feeling unwell retrying with larger and larger intervals. I'm still in the process of establishing of what I think the "canonical" `curl` invocation, so you may see curl options vary while I do that.

I also fixed a minor clippy lint in tests because I added `--all-targets` to `cargo clippy` on CI it started to detect new lints.

I tested the pre-releases with `1.0.0-rc.1` version and it didn't work because the regex didn't include the `.1` part, now it's fixed.

Here are release and a pre-release: [link](https://github.com/Veetaha/marker/releases),

Script for installing a release version should just work:
```
curl --location --silent --fail --show-error --retry 5 --retry-connrefused \
    https://raw.githubusercontent.com/Veetaha/marker/v0.3.0/scripts/release/install.sh | bash
```
Script for installing a pre-release version should just work as well:
```
curl --location --silent --fail --show-error --retry 5 --retry-connrefused \
    https://raw.githubusercontent.com/Veetaha/marker/v0.4.0-rc.123/scripts/release/install.sh | bash
```

---

Don't forget to add `bash-test` to the required checks for the PR, and double-check you did all the requirements mentioned after the following sentence in the PR description.

> "As a result of this PR the following changes to marker infrastructure need to be done manually and only once"

xFrednet commented:
Everything looks good to me, thank you so much for all the work you put into this!
